File: Web Game/src/systems/GameManager.js
/**
 * GameManager - Singleton for global game state
 * Manages player data, progression, and scoring
 */

class GameManager {
    constructor() {
        console.log('GameManager constructor called');
        
        if (GameManager.instance) {
            console.log('GameManager instance already exists, returning');
            return GameManager.instance;
        }

        // Level Data
        this.currentLevel = this.loadFromStorage('currentLevel', 1);
        this.score = 0;
        this.waterAmount = 0;
        this.waterNeeded = 100;
        this.cumulativeScore = this.loadFromStorage('cumulativeScore', 0);

        // Settings
        this.debugMode = true;
        this.soundEnabled = this.loadFromStorage('soundEnabled', true);

        GameManager.instance = this;
        console.log('GameManager initialized successfully');
    }

    /**
     * Reset level data for new attempt
     */
    resetLevel() {
        this.score = 0;
        this.waterAmount = 0;

        if (this.debugMode) {
            console.log(`Level ${this.currentLevel} reset. Score: ${this.score}, Water: ${this.waterAmount}/${this.waterNeeded}`);
        }
    }

    /**
     * Complete the current level and advance
     */
    completeLevel() {
        this.cumulativeScore += this.score;
        this.currentLevel++;

        this.saveToStorage('currentLevel', this.currentLevel);
        this.saveToStorage('cumulativeScore', this.cumulativeScore);

        if (this.debugMode) {
            console.log(`Level completed! Score: ${this.score}, Cumulative: ${this.cumulativeScore}`);
        }
    }

    /**
     * Add water particles to current level
     */
    addWater(amount = 1) {
        this.waterAmount += amount;

        if (this.debugMode) {
            console.log(`Water added: ${amount}. Total: ${this.waterAmount}/${this.waterNeeded}`);
        }
    }

    /**
     * Check if level is complete
     */
    isLevelComplete() {
        return this.waterAmount >= this.waterNeeded;
    }

    /**
     * Check if milestone reward is earned
     */
    checkMilestone(threshold) {
        return this.cumulativeScore >= threshold;
    }

    /**
     * Save data to localStorage
     */
    saveToStorage(key, value) {
        try {
            localStorage.setItem(`cw_${key}`, JSON.stringify(value));
        } catch (e) {
            console.warn('Failed to save to localStorage:', e);
        }
    }

    /**
     * Load data from localStorage
     */
    loadFromStorage(key, defaultValue = null) {
        try {
            const value = localStorage.getItem(`cw_${key}`);
            return value ? JSON.parse(value) : defaultValue;
        } catch (e) {
            console.warn('Failed to load from localStorage:', e);
            return defaultValue;
        }
    }

    /**
     * Reset all progress (for testing)
     */
    resetAllProgress() {
        this.currentLevel = 1;
        this.cumulativeScore = 0;
        this.score = 0;
        this.waterAmount = 0;

        localStorage.removeItem('cw_currentLevel');
        localStorage.removeItem('cw_cumulativeScore');

        if (this.debugMode) {
            console.log('All progress reset.');
        }
    }

    /**
     * Toggle sound
     */
    setSoundEnabled(enabled) {
        this.soundEnabled = enabled;
        this.saveToStorage('soundEnabled', enabled);
    }

    /**
     * Get singleton instance
     */
    static getInstance() {
        if (!GameManager.instance) {
            new GameManager();
        }
        return GameManager.instance;
    }
}

// Create and export singleton instance
const gameManager = GameManager.getInstance();

File: Unity Project/Assets/Scripts/Managers/GameManager.cs
using UnityEngine;

/// <summary>
/// Singleton GameManager that handles global game state and player data persistence.
/// Manages score, cumulative water, level progression, and data that persists across scenes.
/// </summary>
public class GameManager : MonoBehaviour
{
    public static GameManager Instance { get; private set; }

    [Header("Level Data")]
    public int currentLevel = 1;
    public int score = 0;
    public int waterAmount = 0;
    public int waterNeeded = 100;
    public int cumulativeScore = 0;

    [Header("Settings")]
    [SerializeField] private bool debugMode = false;

    private void Awake()
    {
        // Implement singleton pattern
        if (Instance != null && Instance != this)
        {
            Destroy(gameObject);
            return;
        }

        Instance = this;
        DontDestroyOnLoad(gameObject);

        // Load player data from persistent storage
        LoadPlayerData();
    }

    private void OnDestroy()
    {
        if (Instance == this)
        {
            SavePlayerData();
        }
    }

    /// <summary>
    /// Reset level data for a new level attempt
    /// </summary>
    public void ResetLevel()
    {
        score = 0;
        waterAmount = 0;
        
        if (debugMode)
            Debug.Log($"Level {currentLevel} reset. Score: {score}, Water: {waterAmount}/{waterNeeded}");
    }

    /// <summary>
    /// Complete the current level and advance to next
    /// </summary>
    public void CompleteLevel()
    {
        cumulativeScore += score;
        currentLevel++;
        
        SavePlayerData();
        
        if (debugMode)
            Debug.Log($"Level completed! Score: {score}, Cumulative: {cumulativeScore}");
    }

    /// <summary>
    /// Add water particles to the current level
    /// </summary>
    public void AddWater(int amount = 1)
    {
        waterAmount += amount;
        
        if (debugMode)
            Debug.Log($"Water added: {amount}. Total: {waterAmount}/{waterNeeded}");
    }

    /// <summary>
    /// Check if level is complete
    /// </summary>
    public bool IsLevelComplete()
    {
        return waterAmount >= waterNeeded;
    }

    /// <summary>
    /// Check if milestone reward is earned
    /// </summary>
    public bool CheckMilestone(int milestoneThreshold)
    {
        return cumulativeScore >= milestoneThreshold;
    }

    /// <summary>
    /// Save player data to PlayerPrefs
    /// </summary>
    private void SavePlayerData()
    {
        PlayerPrefs.SetInt("CurrentLevel", currentLevel);
        PlayerPrefs.SetInt("CumulativeScore", cumulativeScore);
        PlayerPrefs.Save();
        
        if (debugMode)
            Debug.Log("Player data saved.");
    }

    /// <summary>
    /// Load player data from PlayerPrefs
    /// </summary>
    private void LoadPlayerData()
    {
        currentLevel = PlayerPrefs.GetInt("CurrentLevel", 1);
        cumulativeScore = PlayerPrefs.GetInt("CumulativeScore", 0);
        
        if (debugMode)
            Debug.Log($"Player data loaded. Level: {currentLevel}, Cumulative Score: {cumulativeScore}");
    }

    /// <summary>
    /// Reset all player progress (for testing)
    /// </summary>
    public void ResetAllProgress()
    {
        currentLevel = 1;
        cumulativeScore = 0;
        score = 0;
        waterAmount = 0;
        
        PlayerPrefs.DeleteAll();
        
        if (debugMode)
            Debug.Log("All progress reset.");
    }
}
